pkg/common/watchdog_test.go

```diff
@@ -0,0 +1,128 @@
+package common_test
+
+import (
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/waterfall/pkg/common"
+	"github.com/stretchr/testify/assert"
+)
+
+func testSetup(t *testing.T) *common.Watchdog {
+	t.Helper()
+	w := common.NewWatchdog(2*time.Second, func() {})
+
+	t.Cleanup(func() {
+		w.Close()
+	})
+	return w
+}
+
+func TestWatchdog_Start(t *testing.T) {
+	w := testSetup(t)
+	terminated := w.Start()
+	select {
+	case <-terminated:
+		t.Fatal("Should terminated only after call Close")
+	default:
+		assert.True(t, true)
+	}
+}
+
+func TestWatchdog_Close(t *testing.T) {
+	w := testSetup(t)
+	var terminated chan struct{}
+	terminated = w.Start()
+
+	select {
+	case <-terminated:
+		t.Fatal("Should terminated after call Close")
+	default:
+	}
+
+	w.Close()
+	assert.Empty(t, <-terminated)
+}
+
+func TestWatchdog_Notify(t *testing.T) {
+	w := testSetup(t)
+	w.Start()
+	// Take care: Notify is blocking and this lines running sequential.
+	// The previous implementation was blocking too. I'll leave it like this
+	assert.True(t, w.Notify())
+	assert.True(t, w.Notify())
+	assert.True(t, w.Notify())
+
+	w.Close()
+	assert.False(t, w.Notify())
+	assert.False(t, w.Notify())
+
+	w.Close()
+	assert.False(t, w.Notify())
+	assert.False(t, w.Notify())
+}
+
+func TestWatchdog_Close_before_Start(t *testing.T) {
+	// You can close the watchdog before starting it. I don't know if this is a feature ;-)
+	w := testSetup(t)
+	w.Close()
+	assert.Empty(t, <-w.Start())
+}
+
+func TestWatchdog_Multiple_Start(t *testing.T) {
+	// You can start Watchdog several times.
+	// I think this is a bug, but I didn't want to change the interface.
+	// I would prefer make start private and start the watchdog immediately upon creation to prevent this.
```

Comment:
I agree that it sounds like a bug, that's why the interface of the previous implementation was designed slightly differently (to now allow calling `Start()` multiple times). Could we probably get the previous interface, please? 🙂 (less possibilities to introduce a bug)